### PR TITLE
Bugfix - conversions in _copy_and_shift_params

### PR DESCRIPTION
**Pull request by vincentmr**
### Before submitting

Please complete the following checklist when submitting a PR:

- [x] All new features must include a unit test.
      If you've fixed a bug or added code that should be tested, add a test to the
      test directory!

- [x] All new functions and code must be clearly commented and documented.
      If you do make documentation changes, make sure that the docs build and
      render correctly by running `make docs`.

- [x] Ensure that the test suite passes, by running `make test`.

- [x] Add a new entry to the `doc/releases/changelog-dev.md` file, summarizing the
      change, and including a link back to the PR.

- [x] The PennyLane source code conforms to
      [PEP8 standards](https://www.python.org/dev/peps/pep-0008/).
      We check all of our code against [Pylint](https://www.pylint.org/).
      To lint modified files, simply `pip install pylint`, and then
      run `pylint pennylane/path/to/file.py`.

When all the above are checked, delete everything above the dashed
line and fill in the pull request template.

------------------------------------------------------------------------------------------------------------

**Context:**
[PR 4415](https://github.com/PennyLaneAI/pennylane/pull/4415) breaks the Lightning-Kokkos CI where we test parameter shift with integer parameters.

**Description of the Change:**
Skip cast/convert when the parameter is an integral type.

**Benefits:**
Bug fix.

**Possible Drawbacks:**
Slight slowdown because of the check. It should be unnoticeable in virtually all cases.

**Related GitHub Issues:**
[PR 4415](https://github.com/PennyLaneAI/pennylane/pull/4415)

### CODE REVIEW

File: pennylane/gradients/general_shift_rules.py [hunk 1]
```diff
@@ -397,9 +397,9 @@ def _copy_and_shift_params(tape, indices, shifts, multipliers, cast=False):
         # Shift copied parameter
         new_params = list(op.data)
```
Comment (albi3ro) [inline pennylane/gradients/general_shift_rules.py:398]: We're needing to cast `new_params` to float64 too.

Comment (vincentmr) [reply pennylane/gradients/general_shift_rules.py:398]: If I cast it with say
```
new_params[p_idx] = qml.math.cast(new_params[p_idx], np.float64)
```
then I get an error 
```
ImportError: autoray couldn't find function 'astype' for backend 'builtins'.
```
Do you know what to do with that?

Comment (vincentmr) [reply pennylane/gradients/general_shift_rules.py:398]: Think I found a patch by branching `cast`.

Comment (vincentmr) [reply pennylane/gradients/general_shift_rules.py:398]: My solution was similarly patchy. It breaks some other tests. Playing with yours, I have issues with autograd and complex in other tests.

Comment (albi3ro) [reply pennylane/gradients/general_shift_rules.py:398]: How hard can it be just to convert data to float!

Comment (vincentmr) [reply pennylane/gradients/general_shift_rules.py:399]: Too hard! I instead added a condition to skip casting for integral types.

File: pennylane/math/utils.py
Comment (albi3ro) [inline pennylane/math/utils.py:1]: https://github.com/PennyLaneAI/pennylane/pull/4472

File: tests/gradients/parameter_shift/test_parameter_shift.py [hunk 2]
```diff
@@ -1567,6 +1567,27 @@ def test_single_expectation_value(self, tol):
...
+    )  # integers, zero
+    def test_integer_parameters(self, tol, par):
+        """Test that the gradient of the RY gate matches the exact analytic formula."""
+        dev = qml.device("default.qubit", wires=2)
+
+        with qml.tape.QuantumTape() as tape:
+            qml.RY(par, wires=[0])
+            qml.expval(qml.PauliX(0))
+
+        tape.trainable_params = {0}
```
Comment (albi3ro) [inline tests/gradients/parameter_shift/test_parameter_shift.py:1581]:
```suggestion
        tape = qml.tape.QuantumScript([qml.RY(par, wires=[0])], [qml.expval(qml.PauliX(0))])
        tape.trainable_params = {0}
```

Unless we are actually testing something involving queuing, I prefer to use the "construct on initialization" syntax. No need for a mutable state and a global interpreter lock this way.

**Comment by vincentmr** [comment]
After merging with master, there is one test failing: `tests/interfaces/default_qubit_2_integration/test_jax_qnode_default_qubit_2.py::TestQubitIntegrationHigherOrder::test_hessian[jax-dev12-spsa-False]`.

At first, I could not reproduce on my laptop. I realized that this is because if had `jax==0.4.13 jaxlib==0.4.13` installed. Upgrading to `jax==0.4.14 jaxlib==0.4.14`, the failure is reproducible. I not that the CI before the merge also used `jax==0.4.14 jaxlib==0.4.14`, which suggests there is some interaction between the JAX version and the latest update. 

Since the failure is for SPSA only, could it be related to an RNG update?

File: doc/releases/changelog-dev.md [hunk 3]
```diff
@@ -343,6 +343,9 @@ array([False, False])
 
 <h3>Bug fixes 🐛</h3>
 
+* `_copy_and_shift_params` does not cast or convert integral types, just relying on `+` and `*`'s casting rules in this case.
```
Comment (albi3ro) [inline doc/releases/changelog-dev.md:346]: Is this still the correct descriptor?

Comment (vincentmr) [reply doc/releases/changelog-dev.md:346]: Yes, I would say that's what ended up working (these [lines](https://github.com/PennyLaneAI/pennylane/pull/4477/files#diff-6c7e71f5581ca8693c0844d865d643db342ee0316d0ace61b6f79bfad4b8b572R400-R404)).

File: pennylane/gradients/general_shift_rules.py [hunk 4]
```diff
@@ -396,10 +397,11 @@ def _copy_and_shift_params(tape, indices, shifts, multipliers, cast=False):
...
-        multiplier = qml.math.convert_like(multiplier, new_params[p_idx])
-        multiplier = qml.math.cast_like(multiplier, new_params[p_idx])
-        shift = qml.math.convert_like(shift, new_params[p_idx])
-        shift = qml.math.cast_like(shift, new_params[p_idx])
+        if not isinstance(new_params[p_idx], numbers.Integral):
+            multiplier = qml.math.convert_like(multiplier, new_params[p_idx])
+            multiplier = qml.math.cast_like(multiplier, new_params[p_idx])
+            shift = qml.math.convert_like(shift, new_params[p_idx])
+            shift = qml.math.cast_like(shift, new_params[p_idx])
```
Comment (albi3ro) [inline pennylane/gradients/general_shift_rules.py:404]:
```suggestion
        if qml.math.get_interface(new_params[p_idx]) == "tensorflow":
            multiplier = qml.math.cast_like(multiplier, new_params[p_idx])
            shift = qml.math.cast_like(shift, new_params[p_idx])
        multiplier = qml.math.convert_like(multiplier, new_params[p_idx]) 
        shift = qml.math.convert_like(shift, new_params[p_idx])
```

Can't believe I didn't think of this before.

Comment (vincentmr) [reply pennylane/gradients/general_shift_rules.py:404]: Looks good, but will that work if `qml.math.get_interface(new_params[p_idx]) == "tensorflow"` and `new_params[p_idx]` is an integer? Or does that combination never happen?

Comment (albi3ro) [reply pennylane/gradients/general_shift_rules.py:404]:
```suggestion
        if qml.math.get_interface(new_params[p_idx]) == "tensorflow":
            new_param_p_idx = qml.math.cast(new_params[p_idx], float)
       else:
           new_param_p_idx = new_params[p_idx]
        multiplier = qml.math.convert_like(multiplier, new_param_p_idx ) 
        shift = qml.math.convert_like(shift, new_param_p_idx )
```

Maybe?

**Review by albi3ro — APPROVED** [review]
Thanks. Sorry for causing this problem.

File: pennylane/gradients/general_shift_rules.py (hunk 4 above)
Comment (josh146) [reply pennylane/gradients/general_shift_rules.py:404]: Just some context, but in the original implementation we tried to ensure we weren't causing TF warning due to incompatible casting, as we were getting lots of user bugs. cc @trbromley

Comment (albi3ro) [reply pennylane/gradients/general_shift_rules.py:404]: https://github.com/PennyLaneAI/pennylane/issues/3787

**Review by AmintorDusko — APPROVED** [review]
Nice job here, @vincentmr. Thank you for that!

**Review by mudit2812 — APPROVED** [review]
Thanks!

File: pennylane/gradients/general_shift_rules.py (hunk 4 above)
Comment (trbromley) [reply pennylane/gradients/general_shift_rules.py:404]: To confirm @josh146's point, are we making sure that random TF warnings are not being encountered by PL users (with the TF interface)?

Comment (vincentmr) [reply pennylane/gradients/general_shift_rules.py:404]: The warning above is reproducible with
```
pytest -s tests/interfaces/default_qubit_2_integration/test_tensorflow_default_qubit_2.py -k 'TestTensorflowExecuteIntegration and test_ragged_differentiation'
```
 but I don't think it's a result of this PR. It shows in `6b149a19cd1c7100` and `v0.31.0`.

**Pull request merged**